### PR TITLE
mock: allow pre-creating users in chroot

### PR DESCRIPTION
**Pull request by praiskup**

### CODE REVIEW

File: mock/py/mockbuild/buildroot.py [hunk 1]
```diff
@@ -466,6 +468,9 @@ def _fixup_build_user(self):
     @traceLog()
     @noop_in_bootstrap
     def _make_build_user(self):
+        self.shadow_utils.create_user("pesign", copy_from_host=True)
+        self.shadow_utils.create_user("random-user")
```
Comment (praiskup) [inline mock/py/mockbuild/buildroot.py:472]: This part likely needs to be done elsewhere and needs to be configurable!  This is just a dummy preview for experiments.

**Comment by belegdol** [comment]
This fixes #1091 without having to set up ACLs. `/run/pesign` on the host remains owned by `pesign:pesign` even after installing `pesign` in the chroot.

**Comment by xsuchy** [comment]
@praiskup please check buildroot._make_build_user() for historical knowledge. Like deleting the user first, because it may exists from previous run. And fixing the uid.

**Comment by praiskup** [comment]
Yes, I've seen the history - in this case, I think we can afford to keep it simple, at least at the beginning;  if this logic fails, it fails for misconfiguration (which is not very likely because we do this only when the chroot is first initialized).

**Comment by praiskup** [comment]
> @praiskup please check buildroot._make_build_user() for historical knowledge. Like deleting the user first, because it may exists from previous run. And fixing the uid.

I changed my opinion about this, and I agree that removing the user first is better (makes me more confident that UID/GIDs are aligned).

While doing this quite complex feature, I also propose changing the default way of handling `mockbuilder` (user) and `mock` (group) in-chroot -> so we use the same principle in both cases.

Ready for the review.

**Comment by belegdol** [comment]
I tested the updated code and it works as expected, pesign works correctly.

File: mock/mock.spec [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 Summary: Builds packages inside chroots
 Name: mock
-Version: 4.1
+Version: 4.1.post
```
Comment (xsuchy) [inline mock/mock.spec:12]: ???

File: mock/py/mockbuild/buildroot.py [hunk 3]
```diff
@@ -452,42 +454,26 @@ def _init_pkg_management(self):
...
+        # Resolving the 'mockbuilder' user
         excluded = [self.make_chroot_path(self.homedir, path)
                     for path in self.config['exclude_from_homedir_cleanup']] + \
             self.mounts.get_mountpoints()
         file_util.rmtree(self.make_chroot_path(self.homedir),
                     selinux=self.selinux, exclude=excluded)
 
-        # ok for these two to fail
-        if self.config['clean']:
-            self.doChroot(['/usr/sbin/userdel', '-r', '-f', self.chrootuser],
```
Comment (xsuchy) [inline mock/py/mockbuild/buildroot.py:477]: The -r is lost with the new code.

**Comment by xsuchy** [comment]
> There's one different thing.  We we don't do `userdel -r` anymore for
mockbuilder.  This shouldn't be needed, because when user is really
re-created (prebuild == True), we already did `_cleanup_homedir()`.
The other thing is that before we `useradd`, we already have `/builddir`
pre-created -> so useradd doesn't create `.bashrc`, etc.  We can iterate
on this, but it doesn't seem to be Mock's responsibility.

I do not  get this. When you call mock with `--clean` you want to clean /builddir and with `-noclean` you want to preserve the content. Not just on the first run, but on consecutive runs.

File: mock/mock.spec (hunk 2 above)
Comment (praiskup) [reply mock/mock.spec:12]: This is an experiment (likely I should use `post1`) that I don't insist on,
but seems really attractive.

Simply, this way I can signalize that the Version is higher than the one
released in Fedora, always beating the NVR there (even if Release goes up due
to rebuilds) and in mock-core-configs I still can require the version newer
then the one from this PR.

**Comment by praiskup** [comment]
> I do not get this. When you call mock with --clean you want to clean /builddir
> and with -noclean you want to preserve the content. Not just on the first run,
> but on consecutive runs.

I don't think [--no-clean](https://github.com/rpm-software-management/mock/blob/05cfc1185eda51650956b95d34b288cb30665f31/mock/py/mockbuild/buildroot.py#L469-L473) works this way.  That's executed unconditionally (note that `_make_build_user()` is always called for `prebuild == True`).  Which IMO implies that playing with `-r` option is just redundant (those files are already removed).

**Comment by praiskup** [comment]
But honestly, I'm a bit confused, don't ask me what is the effect of `--no-clean` actually.

File: mock/mock.spec (hunk 2 above)
Comment (praiskup) [reply mock/mock.spec]: I moved this experiment into #1138, it was rather disturbing change in this PR.

**Comment by praiskup** [comment]
So, the consequence here is that e.g. manual `/etc/passwd` modifications
are going to be dropped between `mock --shell` calls, because instead of
using `usermod` calls (old behavior), we are now newly re-creating the
from scratch (new behavior).  The new behavior can drop potentially
user-changed `/etc/{passwd,group}` entries.

Per documentation

        -n, --no-clean
               Do not clean chroot before building a package.
        ...
        --buildsrpm
               Build the specified SRPM either from  a  spec  file  and  source
               file/directory  or SCM. The chroot (including the results direc‐
               tory) is cleaned first, unless --no-clean is specified.

So keeping the old behavior for the "non build" use case seems like we
might be overcautious.  That said, the `--no-clean` option of course has
a meaning, but the thing users really want is to not clean up
the installed packages from the previous build (to potentially speed up
the subsequent build(s).

**Comment by praiskup** [comment]
We know that @voxik and @mkoncek are heavy `--no-clean` users.  Folks, would you mind testing this not-yet-merged variant of mock?  Steps to install this:

    sudo dnf copr enable @mock/mock-pull-requests:pr:1103
    sudo dnf copr update mock mock-core-configs
    mock --scrub=all -r <your chroot>
    ... test the use-case you typically have ...

**Comment by mkoncek** [comment]
Sure, can start tomorrow.
Just to be clear, my whole usage of `--no-clean` is:
* Build newer version of a dependency
* Install newer version of a dependency in mock
* Build newer version of a dependent package with `--no-clean`

I hope this is how it is intended to be used.

**Comment by hobbes1069** [comment]
> Sure, can start tomorrow. Just to be clear, my whole usage of `--no-clean` is:
> 
> * Build newer version of a dependency
> * Install newer version of a dependency in mock
> * Build newer version of a dependent package with `--no-clean`

I use `--no-clean` via `fedpkg mockbuild` when I'm doing test builds of new versions of packages and the build fails for some reason. No point in reinstalling the whole buildroot.

**Comment by praiskup** [comment]
Thank you for responding here!

@hobbes1069 wrote
> I use --no-clean via fedpkg mockbuild when I'm doing test builds of new versions of packages and the build fails for some reason. No point in reinstalling the whole buildroot.

Indeed, this behavior of --no-clean shouldn't be affected by this PR.

@fweimer-rh wrote
> The --shell option with the implied --no-clean still overwrites /etc/dnf/dnf.conf even before this change, which I found terribly confusing at one point.

Good point.  There already are other things that Mock sets up for every run no matter `--{,no-}clean`.  So this isn't precedent at least.

> But I think rewriting /etc/passwd is different because it reverts the
> the actions of postinst scripts of already-installed packages. I think
> this means that just re-running make in the otherwise unchanged
> buildroot re-entered via --shell can produce unexpected results. That
> can't be good?

Well, in theory at least (we may e.g. do `mock --shell` => `mock --install missinb-builddep` => `mock --shell` again).  But the original motivation was different dcad1667d16c1ee5c0be203f7253447fbe8dcdfe (it's not obvious, but the modified `else:` branch calling `_fixup_build_user` basically relates to `--shell`, not `--rebuild`, which btw did re-create the user even before, since 7814ebd4d074d5434b7984a28508d2a422aa156f).  Note mock always does other things like "chown -R <homedir>" so changing UID actually has some real effect.

**Review by xsuchy — APPROVED** [review]

**Pull request merged**